FileExistsMode.REPLACE not working

In version 6.0.3

I'm new to spring integration

FileExistsMode.REPLACE throw exception instead it changes the original file.
I don't know if this is a bug.

![image](https://user-images.githubusercontent.com/46040940/221626185-ef026aac-63ca-40dd-8dfe-7104aff62886.png)

Just came along this scenario in my project .

We recently switched to use IntegrationFlow with FileExistsMode.FAIL in the first step.
After a couple of days business came along and requested to set FileExistsMode.REPLACE instead.

As of now we are facing exactly the same issue.

We do have enabled useTemporaryFilename.
- The SFTP Integration is writing the filename.ext.writing file to SFTP server. ✅ 
- After writing is finished the file is rename to filename.ext ✅ 
- We need to write a file with the same filename again.
- The SFTP Integration is writing the filename.ext.writing file to SFTP server. ✅ 
- After writing is finished the file is rename to filename.ext 🛑 

So in the end the old  filename.ext named file is still on SFTP. 
Next to it there is the filename.ext.writing file with the new content.

And the log says:

`Caused by: org.springframework.messaging.MessagingException: Failed to write to '/develop/export/XXXXXXX.json.writing' while uploading the file
at org.springframework.integration.file.remote.RemoteFileTemplate.sendFileToRemoteDirectory(RemoteFileTemplate.java:573)
at org.springframework.integration.file.remote.RemoteFileTemplate.doSend(RemoteFileTemplate.java:353)
... 74 common frames omitted
Caused by: org.apache.sshd.sftp.common.SftpException: Failure
at org.apache.sshd.sftp.client.impl.AbstractSftpClient.throwStatusException(AbstractSftpClient.java:217)
at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkResponseStatus(AbstractSftpClient.java:212)
at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkResponseStatus(AbstractSftpClient.java:186)
at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkCommandStatus(AbstractSftpClient.java:164)
at org.apache.sshd.sftp.client.impl.AbstractSftpClient.rename(AbstractSftpClient.java:647)
at org.apache.sshd.sftp.client.SftpClient.rename(SftpClient.java:617)
at org.springframework.integration.sftp.session.SftpSession.rename(SftpSession.java:165)
at org.springframework.integration.file.remote.session.CachingSessionFactory$CachedSession.rename(CachingSessionFactory.java:252)
at org.springframework.integration.file.remote.RemoteFileTemplate.doSend(RemoteFileTemplate.java:606)
at org.springframework.integration.file.remote.RemoteFileTemplate.sendFileToRemoteDirectory(RemoteFileTemplate.java:570)
... 75 common frames omitted`

We do have all the rights we need on SFTP Server to perform 

Are we missing any configuration? SFTP Documentation tells me to remove the target file before renaming. I thought SFTP Integration would handle this for me?

--- a/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java
+++ b/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2002-2022 the original author or authors.
+ * Copyright 2002-2023 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -65,7 +65,18 @@ public class SftpSession implements Session<SftpClient.DirEntry> {
 
 	@Override
 	public boolean remove(String path) throws IOException {
-		this.sftpClient.remove(path);
+		try {
+			this.sftpClient.remove(path);
+		}
+		catch (SftpException sftpEx) {
+			if (SftpConstants.SSH_FX_NO_SUCH_FILE == sftpEx.getStatus()) {
+				return false;
+			}
+			else {
+				throw sftpEx;
+			}
+		}
+
 		return true;
 	}
 
@@ -174,19 +185,8 @@ public class SftpSession implements Session<SftpClient.DirEntry> {
 			this.sftpClient.rename(pathFrom, pathTo, SftpClient.CopyMode.Overwrite);
 		}
 		else {
-			try {
-				this.sftpClient.rename(pathFrom, pathTo);
-			}
-			catch (SftpException sftpex) {
-				if (SftpConstants.SSH_FX_FILE_ALREADY_EXISTS == sftpex.getStatus()) {
-					remove(pathTo);
-					// attempt to rename again
-					this.sftpClient.rename(pathFrom, pathTo);
-				}
-				else {
-					throw sftpex;
-				}
-			}
+			remove(pathTo);
+			this.sftpClient.rename(pathFrom, pathTo);
 		}
 	}
 

